File: app/src/main/java/com/example/trigger/https/CertificateFetchTask.java
package com.example.trigger.https;

import java.net.URL;
import java.security.cert.Certificate;
import javax.net.ssl.HttpsURLConnection;

import android.os.AsyncTask;

import com.example.trigger.Log;


public class CertificateFetchTask extends AsyncTask<Object, Void, CertificateFetchTask.Result> {
    private OnTaskCompleted listener;

    public interface OnTaskCompleted {
        void onCertificateFetchTaskCompleted(Result result);
    }

    public static class Result {
        String error;
        Certificate certificate;

        Result(Certificate certificate, String error) {
            this.certificate = certificate;
            this.error = error;
        }
    }

    public CertificateFetchTask(CertificateFetchTask.OnTaskCompleted listener){
        this.listener = listener;
    }

    @Override
    protected Result doInBackground(Object... params) {
        if (params.length != 1) {
            Log.e("CertificateFetchTask", "Unexpected number of params.");
            return new Result(null, "Internal Error");
        }

        try {
            URL url = new URL((String) params[0]);

            // try to establish TLS session only
            int port = (url.getPort() > 0) ? url.getPort() : url.getDefaultPort();
            url = new URL("https", url.getHost(), port, "");

            // disable all certification checks
            HttpsTools.disableDefaultHostnameVerifier();
            HttpsTools.disableDefaultCertificateValidation();

            HttpsURLConnection con = (HttpsURLConnection) url.openConnection();
            con.setConnectTimeout(2000);

            con.connect();
            Certificate[] certificates = con.getServerCertificates();
            con.disconnect();

            if (certificates.length == 0) {
                return new Result(null, "No certificate found.");
            } else {
                return new Result(certificates[0], "");
            }
        } catch (Exception e) {
            return new Result(null, e.toString());
        }
    }

    protected void onPostExecute(Result result) {
        listener.onCertificateFetchTaskCompleted(result);
    }
}

File: app/src/main/kotlin/app/trigger/https/CertificateFetchTask.kt
/*
* Copyright (C) 2025 The Trigger Contributors
* SPDX-License-Identifier: GPL-3.0-or-later
*/

package app.trigger.https

import android.os.AsyncTask
import app.trigger.Log
import java.lang.Exception
import java.net.URL
import java.security.cert.Certificate
import javax.net.ssl.HttpsURLConnection

class CertificateFetchTask(private val listener: OnTaskCompleted) : AsyncTask<Any?, Void?, CertificateFetchTask.Result>() {
    interface OnTaskCompleted {
        fun onCertificateFetchTaskCompleted(result: Result)
    }

    class Result internal constructor(var certificate: Certificate?, var error: String)

    override fun doInBackground(vararg params: Any?): Result {
        if (params.size != 1) {
            Log.e(TAG, "Unexpected number of params.")
            return Result(null, "Internal Error")
        }
        return try {
            var url = URL(params[0] as String)

            // try to establish TLS session only
            val port = if (url.port > 0) url.port else url.defaultPort
            url = URL("https", url.host, port, "")

            // disable all certification checks
            HttpsTools.disableDefaultHostnameVerifier()
            HttpsTools.disableDefaultCertificateValidation()
            val con = url.openConnection() as HttpsURLConnection
            con.connectTimeout = 2000
            con.connect()
            val certificates = con.serverCertificates
            con.disconnect()
            if (certificates.size == 0) {
                Result(null, "No certificate found.")
            } else {
                Result(certificates[0], "")
            }
        } catch (e: Exception) {
            Result(null, e.toString())
        }
    }

    override fun onPostExecute(result: Result) {
        listener.onCertificateFetchTaskCompleted(result)
    }

    companion object {
        const val TAG = "CertificateFetchTask"
    }
}